src/Extension/AttributeExtension.php

```diff
@@ -0,0 +1,179 @@
+ */
+final class AttributeExtension extends AbstractExtension implements WithLastModified
+{
+    private array $filters;
+    private array $functions;
+    private array $tests;
+
+    public function __construct(
+        /**
+         * A list of objects or class names defining filters, functions, and tests using PHP attributes.
+         * When passing a class name, it must be available in runtimes.
+         *
+         * @var iterable<object|class-string>
+         */
+        private iterable $objectsOrClasses,
+    ) {
+    }
+
+    public function getLastModified(): int
+    {
+        $lastModified = 0;
+
+        foreach ($this->objectsOrClasses as $objectOrClass) {
+            $r = new \ReflectionClass($objectOrClass);
+            if (is_file($r->getFileName()) && ($extensionTime = filemtime($r->getFileName())) > $lastModified) {
+                $lastModified = $extensionTime;
+            }
+        }
+
+        return $lastModified;
+    }
+
+    public function getFilters(): array
+    {
+        if (!isset($this->filters)) {
+            $this->initFromAttributes();
+        }
+
+        return $this->filters;
+    }
+
+    public function getFunctions(): array
+    {
+        if (!isset($this->functions)) {
+            $this->initFromAttributes();
+        }
+
+        return $this->functions;
+    }
+
+    public function getTests(): array
+    {
+        if (!isset($this->tests)) {
+            $this->initFromAttributes();
+        }
+
+        return $this->tests;
+    }
+
+    private function initFromAttributes()
+    {
+        $filters = $functions = $tests = [];
+
+        foreach ($this->objectsOrClasses as $objectOrClass) {
+            try {
+                $reflectionClass = new \ReflectionClass($objectOrClass);
+            } catch (\ReflectionException $e) {
+                throw new \LogicException(sprintf('"%s" class requires a list of objects or class name, "%s" given.', __CLASS__, get_debug_type($objectOrClass)), 0, $e);
+            }
+
+            foreach ($reflectionClass->getMethods() as $method) {
+                // Filters
+                foreach ($method->getAttributes(AsTwigFilter::class) as $attribute) {
+                    $attribute = $attribute->newInstance();
+
+                    $name = $attribute->name ?? $method->getName();
+                    if (isset($filters[$name])) {
+                        throw new \LogicException(sprintf('Multiple definitions of the "%s" filter.', $name));
+                    }
+
+                    $parameters = $method->getParameters();
+                    $needsEnvironment = isset($parameters[0]) && Environment::class === $parameters[0]->getType()?->getName();
+                    $firstParam = $needsEnvironment ? 1 : 0;
+                    $needsContext = isset($parameters[$firstParam]) && 'context' === $parameters[$firstParam]->getName() && 'array' === $parameters[$firstParam]->getType()?->getName();
```

Comment:
this special behavior of parameter named `$context` needs to be documented on the attributes.

Reply:
I added the doc and example in the attribute class header.

Reply:
The [Symfony `serialize` filter](https://github.com/symfony/symfony/blob/2d88c405ba0de17aebcf5df3de79b12f3fdb297a/src/Symfony/Bridge/Twig/Extension/SerializerRuntime.php#L29) accepts an `array $context` as 3rd argument. This will not conflict but this means developers could have an unexpected behavior if they don't know about this special argument handling.